### PR TITLE
[DESK][SHELL32] Add desktop icon change feature

### PR DESCRIPTION
**Pull request by binarymaster**
# Purpose

- [SHELL32] Add missing strings for desktop icons
  - Reuse existing translations where possible.
- [DESK] Add desktop icon change feature
  - All functions implemented, except no support for themes.
  - Also fix listview height to avoid vertical scrollbar in Windows.

JIRA issue: [CORE-8427](https://jira.reactos.org/browse/CORE-8427)

Tested and confirmed working in Windows XP and 2003.

# Demo

https://user-images.githubusercontent.com/578406/196244458-98e68b08-52e4-411a-8cd0-fe6d4c235d9e.mp4

### CODE REVIEW

File: dll/cpl/desk/background.c [hunk 1]
```diff
@@ -1146,19 +1146,15 @@ static VOID
...
-    ZeroMemory(&psh, sizeof(psh));
-    psh.dwSize = sizeof(psh);
     psh.dwFlags = PSH_NOAPPLYNOW;
     psh.hwndParent = GetParent(hwndDlg);
     psh.hInstance = hApplet;
     psh.pszCaption = MAKEINTRESOURCE(IDS_DESKTOP_ITEMS);
     psh.phpage = hpsp;
 
-    ZeroMemory(&psp, sizeof(psp));
-    psp.dwSize = sizeof(psp);
```
Comment (binarymaster) [inline dll/cpl/desk/background.c:1161]: Refactored these two thanks to Mark's comment: https://github.com/reactos/reactos/pull/4766#discussion_r992488804

File: dll/cpl/desk/desktop.c [hunk 2]
```diff
@@ -222,9 +335,85 @@ DesktopOnInitDialog(IN HWND hwndDlg, IN PDESKTOP_DATA pData)
...
+
+        lvitem.mask = LVIF_TEXT | LVIF_IMAGE | LVIF_PARAM;
+        lvitem.iItem = i;
+        lvitem.iSubItem = 0;
+        lvitem.pszText = pData->LocalIcon[i].szTitle;
+        lvitem.lParam = (LPARAM)i;
+
+        if (hIcon)
+        {
+            lvitem.iImage = ImageList_AddIcon(pData->hLocalImageList, hIcon);
```
Comment (binarymaster) [inline dll/cpl/desk/desktop.c:387]: Maybe add a NULL check for `pData->hLocalImageList`.

File: dll/cpl/desk/desktop.c [hunk 3]
```diff
@@ -268,6 +464,71 @@ DesktopPageProc(IN HWND hwndDlg, IN UINT uMsg, IN WPARAM wParam, IN LPARAM lPara
...
+                if (bUpdateIcon)
+                {
+                    HWND hwndList = GetDlgItem(hwndDlg, IDC_ICONS_LISTVIEW);
+                    HICON hIcon;
+
+                    hIcon = GetIconFromLocation(pData->LocalIcon[i].szPath);
+
+                    if (hIcon)
+                    {
+                        ImageList_ReplaceIcon(pData->hLocalImageList, i, hIcon);
```
Comment (binarymaster) [inline dll/cpl/desk/desktop.c:503]: And here too.

File: dll/cpl/desk/desktop.c [hunk 4]
```diff
@@ -24,11 +36,30 @@ struct
...
+    LPCTSTR CLSID;
+    UINT Title;
+    LPCTSTR Icon;
+} IconChange[NUM_CHANGE_ICONS] = {
+    {TEXT("{20D04FE0-3AEA-1069-A2D8-08002B30309D}"), IDS_TITLE_MYCOMP, NULL},         /* My Computer */
+    {TEXT("{450D8FBA-AD25-11D0-98A8-0800361B1103}"), IDS_PERSONAL, NULL},             /* My Documents */
+    {TEXT("{208D2C60-3AEA-1069-A2D7-08002B30309D}"), IDS_TITLE_MYNET, NULL},          /* My Network Places */
+    {TEXT("{645FF040-5081-101B-9F08-00AA002F954E}"), IDS_TITLE_BIN_1, TEXT("Full")},  /* Recycle Bin (full) */
+    {TEXT("{645FF040-5081-101B-9F08-00AA002F954E}"), IDS_TITLE_BIN_0, TEXT("Empty")}, /* Recycle Bin (empty) */
+};
```
Comment (learn-more) [inline dll/cpl/desk/desktop.c:55]: TBH It doesn't make sense to write new code with TCHAR

Comment (binarymaster) [reply dll/cpl/desk/desktop.c:55]: I expected these comments as I saw them in other PR :smile: I want to keep consistency with existing code in this file.

File: dll/cpl/desk/desktop.c [hunk 5]
```diff
@@ -102,6 +174,16 @@ SaveDesktopSettings(PDESKTOP_DATA pData)
         pData->optIcons[i].bHideNewStart = pData->bLocalHideIcon[i];
     }
 
+    for (i = 0; i < _countof(IconChange); i++)
+    {
+        if (pData->bLocalIconChanged[i])
+            pData->bIconChanged[i] = TRUE;
+        else
+            continue;
```
[the remark this answers is not in the document]
Comment (binarymaster) [reply dll/cpl/desk/desktop.c]: Improved.

File: dll/cpl/desk/desk.h [hunk 6]
```diff
@@ -81,24 +81,42 @@ ULONG __cdecl DbgPrint(PCCH Format,...);
 #define MAX_DESK_PAGES        32
 #define NUM_SPECTRUM_BITMAPS  3
 
-/* This number must match DesktopIcons array size */
-#define NUM_DESKTOP_ICONS  4
+#define NUM_DESKTOP_ICONS  4  /* DesktopIcons array size */
+#define NUM_CHANGE_ICONS   5  /* IconChange array size */
```
Comment (HBelusca) [inline dll/cpl/desk/desk.h:85]: What's the difference between the two?

File: dll/cpl/desk/desktop.c [hunk 7]
```diff
@@ -7,7 +7,19 @@
...
+
+/* From shresdef.h */
+#define FCIDM_DESKBROWSER_REFRESH    0xA220
+
+#define IDS_TITLE_MYCOMP  30386
+#define IDS_TITLE_MYNET   30387
+#define IDS_TITLE_BIN_1   30388
+#define IDS_TITLE_BIN_0   30389
+
+/* Workaround */
```
Comment (HBelusca) [inline dll/cpl/desk/desktop.c:21]: Workaround for? (should be mentioned in the comment)

File: dll/cpl/desk/desktop.c [hunk 8]
```diff
@@ -81,6 +112,47 @@ InitDesktopSettings(PDESKTOP_DATA pData)
...
+
+        if (SHGetValue(HKEY_CLASSES_ROOT, iconPath, IconChange[i].Icon, &dwType,
+                       &szData, &cbData) == ERROR_SUCCESS &&
+            (dwType == REG_SZ || dwType == REG_EXPAND_SZ))
+        {
+            StringCchCopy(pData->DefIcon[i].szPath, _countof(pData->DefIcon[i].szPath), szData);
+        }
+
+        /* Emergency fallback */
+        if (lstrlen(pData->DefIcon[i].szPath) == 0)
```
Comment (HBelusca) [inline dll/cpl/desk/desktop.c:150]:
```suggestion
        if (!*(pData->DefIcon[i].szPath))
```

File: dll/cpl/desk/desktop.c [hunk 9]
```diff
@@ -81,6 +112,47 @@ InitDesktopSettings(PDESKTOP_DATA pData)
...
+            StringCchCopy(pData->DefIcon[i].szPath, _countof(pData->DefIcon[i].szPath), szFallbackIcon);
+
+        if (lstrlen(pData->Icon[i].szPath) == 0)
```
Comment (HBelusca) [inline dll/cpl/desk/desktop.c:153]:
```suggestion
        if (!*(pData->Icon[i].szPath))
```

File: dll/cpl/desk/desktop.c [hunk 10]
```diff
@@ -187,24 +267,55 @@ SetDesktopSettings(PDESKTOP_DATA pData)
...
+        if (!pData->bIconChanged[i])
+            continue;
+
+        StringCchCopy(iconPath, _countof(iconPath), szUserClass);
+        StringCchCat(iconPath, _countof(iconPath), IconChange[i].CLSID);
+        StringCchCat(iconPath, _countof(iconPath), szDefaultIcon);
+
+        SHSetValue(HKEY_CURRENT_USER, iconPath, IconChange[i].Icon, dwType,
+                   pData->Icon[i].szPath, sizeof(pData->Icon[i].szPath));
+        if (IconChange[i].Title == IDS_TITLE_BIN_0)
```
Comment (HBelusca) [inline dll/cpl/desk/desktop.c:284]: Just to confirm, that `.Title` contains here a numerical ID?

File: dll/cpl/desk/desktop.c [hunk 11]
```diff
@@ -222,9 +333,86 @@ DesktopOnInitDialog(IN HWND hwndDlg, IN PDESKTOP_DATA pData)
...
+        if (hIcon)
+        {
+            if (pData->hLocalImageList)
+                lvitem.iImage = ImageList_AddIcon(pData->hLocalImageList, hIcon);
+            DestroyIcon(hIcon);
+        }
+
+        if (ListView_InsertItem(hwndList, &lvitem) >= 0)
+        {
+            if (i > 0)
```
Comment (HBelusca) [inline dll/cpl/desk/desktop.c:392]: Why not make this test + continue outside of the `if ListView_InsertItem` ?

File: dll/win32/shell32/lang/en-US.rc [hunk 12]
```diff
@@ -980,6 +980,11 @@ BEGIN
     IDS_OBJECTS "%d Objects"
     IDS_OBJECTS_SELECTED "%d Objects Selected"
 
+    IDS_TITLE_MYCOMP "My Computer"
```
Comment (HBelusca) [inline dll/win32/shell32/lang/en-US.rc:983]: Where do these translations come from?

File: dll/cpl/desk/desktop.c (hunk 11 above)
Comment (binarymaster) [reply dll/cpl/desk/desktop.c]: We need to add all items to the list. Now I have restructured this a bit.

File: dll/cpl/desk/desktop.c (hunk 10 above)
Comment (binarymaster) [reply dll/cpl/desk/desktop.c:289]: Yes.

File: dll/cpl/desk/desktop.c (hunk 7 above)
Comment (binarymaster) [reply dll/cpl/desk/desktop.c]: Added comment.

File: dll/cpl/desk/desk.h (hunk 6 above)
Comment (binarymaster) [reply dll/cpl/desk/desk.h:85]: First is for show/hide checkboxes, second is for the icons in listview.

File: dll/win32/shell32/lang/en-US.rc (hunk 12 above)
Comment (binarymaster) [reply dll/win32/shell32/lang/en-US.rc:983]: Translation comes from existing strings (`IDS_MYCOMPUTER`, `IDS_NETWORKPLACE`, `IDS_RECYCLEBIN_FOLDER_NAME`).

File: dll/cpl/desk/desktop.c (hunk 8 above)
Comment (binarymaster) [reply dll/cpl/desk/desktop.c:155]: `szPath` is an array, not a pointer to string. I think `if (!pData->DefIcon[i].szPath[0])` would be more correct in this case, however currently it's more readable and easy to understand. I prefer to leave it as is.

File: dll/cpl/desk/desktop.c [hunk 13]
```diff
@@ -7,7 +7,24 @@
...
+#include <shlwapi.h>
+#include <shellapi.h>
+
+/* From shresdef.h */
+#define FCIDM_DESKBROWSER_REFRESH    0xA220
+
+#define IDS_TITLE_MYCOMP  30386
+#define IDS_TITLE_MYNET   30387
+#define IDS_TITLE_BIN_1   30388
+#define IDS_TITLE_BIN_0   30389
```
Comment (katahiromz) [inline dll/cpl/desk/desktop.c:19]: Move the resource IDs to `resource.h`.

Comment (binarymaster) [reply dll/cpl/desk/desktop.c:19]: These are external resources (from shell32.dll, not desk.cpl), are you sure they needs to be placed there anyway?

File: dll/cpl/desk/desktop.c (hunk 8 above)
Comment (HBelusca) [reply dll/cpl/desk/desktop.c:155]: > szPath is an array, not a pointer to string.

In C these are the same thing.

> I think `if (!pData->DefIcon[i].szPath[0])` would be more correct [...]

It is exactly the same, and maybe more readable.

File: dll/cpl/desk/desktop.c (hunk 13 above)
Comment (katahiromz) [reply dll/cpl/desk/desktop.c:19]: Ah, don't move them.

File: dll/win32/shell32/lang/en-US.rc (hunk 12 above)
Comment (HBelusca) [reply dll/win32/shell32/lang/en-US.rc:983]: In the same file? :thinking: then what about using those existing string IDs?

File: dll/cpl/desk/desktop.c (hunk 13 above)
Comment (katahiromz) [reply dll/cpl/desk/desktop.c:19]: `/* From shell32.dll */`

File: dll/win32/shell32/lang/en-US.rc (hunk 12 above)
Comment (binarymaster) [reply dll/win32/shell32/lang/en-US.rc:983]: Windows has separate resources for these specific things, I'm following it :slightly_smiling_face:

Comment (HBelusca) [reply dll/win32/shell32/lang/en-US.rc:983]: ah, so it duplicates the strings too? SUS but OK....

File: dll/cpl/desk/desktop.c (hunk 13 above)
Comment (binarymaster) [reply dll/cpl/desk/desktop.c:19]: Yes, but there's a comment `/* From shresdef.h */` which is actual for all constants listed below it.

File: dll/cpl/desk/desktop.c [hunk 14]
```diff
@@ -222,9 +338,86 @@ DesktopOnInitDialog(IN HWND hwndDlg, IN PDESKTOP_DATA pData)
...
+        {
+            if (pData->hLocalImageList)
+                lvitem.iImage = ImageList_AddIcon(pData->hLocalImageList, hIcon);
+            DestroyIcon(hIcon);
+        }
+
+        if (ListView_InsertItem(hwndList, &lvitem) < 0)
+            continue;
+
+        if (i > 0)
```
Comment (katahiromz) [inline dll/cpl/desk/desktop.c:398]: What purpose of `if (i > 0) continue;`?

Comment (binarymaster) [reply dll/cpl/desk/desktop.c:398]: The code below sets the item as focused and active. `if (i > 0) continue;` is here to apply these properties only to the first item.

Comment (HBelusca) [reply dll/cpl/desk/desktop.c:398]: Isn't first item index == 0 ?

File: dll/cpl/desk/desktop.c [hunk 15]
```diff
@@ -222,9 +338,86 @@ DesktopOnInitDialog(IN HWND hwndDlg, IN PDESKTOP_DATA pData)
...
+            {
+                /* Fallback to predefined strings */
+                LoadString(GetModuleHandle(TEXT("shell32.dll")),
+                           IconChange[i].Title,
+                           pData->LocalIcon[i].szTitle,
+                           _countof(pData->LocalIcon[i].szTitle));
+            }
+        }
+
+        hIcon = GetIconFromLocation(pData->LocalIcon[i].szPath);
```
Comment (katahiromz) [inline dll/cpl/desk/desktop.c:380]: Another newly-defined function?

File: dll/cpl/desk/desktop.c (hunk 14 above)
Comment (binarymaster) [reply dll/cpl/desk/desktop.c:398]: Yes it is.

File: dll/cpl/desk/desktop.c (hunk 15 above)
Comment (binarymaster) [reply dll/cpl/desk/desktop.c:380]: It's local function defined in this file.

File: dll/cpl/desk/desktop.c (hunk 14 above)
Comment (HBelusca) [reply dll/cpl/desk/desktop.c:398]: So why do we need to skip it?

Comment (binarymaster) [reply dll/cpl/desk/desktop.c:398]: We're skipping everything larger than 0 here, not vice-versa.

File: dll/cpl/desk/desktop.c (hunk 15 above)
Comment (katahiromz) [reply dll/cpl/desk/desktop.c:380]: Looks complicated. Can you move these code to new function?

Comment (HBelusca) [reply dll/cpl/desk/desktop.c:380]: what?

Comment (binarymaster) [reply dll/cpl/desk/desktop.c:380]: It's just initialization of two components, the titles and the icons. Not really complicated. I don't see a good reason to move into function.

File: dll/cpl/desk/desktop.c (hunk 14 above)
Comment (katahiromz) [reply dll/cpl/desk/desktop.c:398]: Why the first item is special?

Comment (katahiromz) [reply dll/cpl/desk/desktop.c:398]: We can do it at outside of this loop.

Comment (binarymaster) [reply dll/cpl/desk/desktop.c:398]: Perhaps we can do it outside the loop, but I based it on this code which does it in loop: https://git.reactos.org/?p=reactos.git;a=blob;f=dll/cpl/inetcpl/security.c;hb=a33719500c71638da03e83e0c9914ec9c9fe3c27#l211

**Review by katahiromz — APPROVED** [review]

**Review by HBelusca — APPROVED** [review]

**Pull request merged**